File: Web/API/FalaCidade.API/Services/NotificationService.cs
using FalaCidade.API.Data;
using FalaCidade.API.Entities;
using Microsoft.EntityFrameworkCore;

namespace FalaCidade.API.Services;

public class NotificationService
{
    private readonly AppDbContext _context;

    public NotificationService(AppDbContext context)
    {
        _context = context;
    }
    public async Task<Notification> CreateAsync(int userId, string message, int? occurrenceId = null)
    {
        var notification = new Notification
        {
            UserId = userId,
            Message = message,
            OccurrenceId = occurrenceId,
            IsRead = false,
            CreatedAt = DateTime.UtcNow
        };

        _context.Notifications.Add(notification);
        await _context.SaveChangesAsync();

        return notification;
    }

    public async Task<IEnumerable<Notification>> GetByUserIdAsync(int userId, bool onlyUnread = false)
    {
        var query = _context.Notifications
            .AsNoTracking()
            .Where(n => n.UserId == userId);

        if (onlyUnread)
        {
            query = query.Where(n => !n.IsRead);
        }

        return await query.OrderByDescending(n => n.CreatedAt).ToListAsync();
    }

    public async Task<bool> MarkAsReadAsync(int notificationId)
    {
        var notification = await _context.Notifications.FindAsync(notificationId);

        if (notification == null) return false;

        notification.IsRead = true;
        await _context.SaveChangesAsync();

        return true;
    }

    public async Task MarkAllAsReadAsync(int userId)
    {
        var unreadNotifications = await _context.Notifications
            .Where(n => n.UserId == userId && !n.IsRead)
            .ToListAsync();

        if (!unreadNotifications.Any()) return;

        foreach (var notification in unreadNotifications)
        {
            notification.IsRead = true;
        }

        await _context.SaveChangesAsync();
    }
}

File: Web/frontend/src/services/notificationService.ts
import api from './api';

export interface Notification {
  id: number;
  occurrenceId: number;
  message: string;
  createdAt: string;
  isRead: boolean;
}

const NotificationService = {
    getAllByUser: async (userId: number): Promise<Notification[]> => {
        const response = await api.get<Notification[]>(`/api/notification/user/${userId}`);
        return response.data;
    },

    getUnread: async (userId: number): Promise<Notification[]> => {
        const response = await api.get<Notification[]>(`/api/notification/user/${userId}/unread`);
        return response.data;
    },

    markAsRead: async (notificationId: number): Promise<void> => {
        await api.patch(`/api/notification/${notificationId}/read`);
    }
};

export default NotificationService;